### ISSUE TITLE
Knex timeout acquiring a connection on strapi start

### ISSUE THREAD

**Issue by puneetverma05** [strapi/strapi#2790]
**Informations**
- **Node.js version**: * v10.15.1
- **NPM version**: * 6.4.1
- **Strapi version**: * 3.0.0-alpha.23
- **Database**: MySQL 8
- **Operating system**: Ubuntu 18.10
- **(Optional) Link to your Project**: 

**What is the current behavior?**
When strarting the project first time with strapi start.
`(node:10022) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?`

**Steps to reproduce the problem**

Create a project with Database MySQL 8.

**What is the expected behavior?**

**Suggested solutions**

**Comment by lauriejim** [comment]
Hello @puneetverma05 !

I need more information.

Can you tell me:
- where you DB is hosted
- if the database test is a success during the application creation
- need the full log of your `strapi start` not just the error

**Comment by puneetverma05** [comment]
**DB is hosted Locally .**
**True** **, Database test is a success during initiazliing the app.**
Here are the logs:
```
$ strapi start
[2019-02-09T18:35:58.962Z] debug ⛔️ Server wasn't able to start properly.
[2019-02-09T18:35:58.963Z] error (hook:bookshelf) takes too long to load
(hook:bookshelf) takes too long to load
```

**After increasing the timeout time from 3000 to 10000 (hook.json) :**

```
strapi start
(node:10963) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/home/projectDir/node_modules/knex/lib/client.js:336:13)
From previous event:
    at Client_MySQL.acquireConnection (/home/projectDir/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/home/projectDir/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/home/projectDir/node_modules/knex/lib/runner.js:33:41)
    at Builder.Target.then (/home/projectDir/node_modules/knex/lib/interface.js:23:43)
    at Child.<anonymous> (/home/projectDir/node_modules/bookshelf/lib/model.js:1118:36)
From previous event:
    at Child.<anonymous> (/home/projectDir/node_modules/bookshelf/lib/model.js:1117:125)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Child.<anonymous> (/home/projectDir/node_modules/bookshelf/lib/model.js:1000:8)
From previous event:
    at Function.addPermission (/home/projectDir/plugins/users-permissions/config/queries/bookshelf.js:152:8)
    at Promise.all.toAdd.map.map.action (/home/projectDir/plugins/users-permissions/services/UsersPermissions.js:293:18)
    at Array.map (<anonymous>)
    at Promise.all.roles.map.role (/home/projectDir/plugins/users-permissions/services/UsersPermissions.js:292:16)
    at Array.map (<anonymous>)
    at Object.updatePermissions (/home/projectDir/plugins/users-permissions/services/UsersPermissions.js:288:15)
(node:10963) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:10963) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[2019-02-09T18:39:56.406Z] warn Bootstrap is taking unusually long to execute its callback 3500 miliseconds).
[2019-02-09T18:39:56.409Z] warn Perhaps you forgot to call it?
```

**Comment by puneetverma05** [comment]
Issue persists when trying with [strapi-docker](https://github.com/strapi/strapi-docker) (on MySQL 8) .
But when using **MySQL 5.6 Container It is working fine** as expected :100:

**Comment by derrickmehaffy** [comment]
I am unable to reproduce this using MariaDB 10.3 (MySQL 8 fork)

**Comment by puneetverma05** [comment]
Was using the mysql community version.
```
db:
    image: mysql:8
    command: --default-authentication-plugin=mysql_native_password
    volumes:
      ...
```

**Comment by derrickmehaffy** [comment]
@puneetverma05 Can you please test using a non-docker version of MySQL 8? (Such as a Virtual machine)

**Comment by puneetverma05** [comment]
@derrickmehaffy  I have tested first on the non-docker version of MySQL8.

**Comment by pdsm** [comment]
I am having the same issue on GCP, any ideas?

**Comment by raphaelmayer** [comment]
I am having a similar issue (MySQL 5.7.19). 
Had no issues until today, when put my bundled react app into the public folder to be served on "/".
Got this exact error, closed and reopened terminal and it worked again. 

However when I went to localhost:1337 (where I serve a bundled react app, which fetches products in compDidMount) I first get a similar error, but after refreshing the browser window it works. Same deal when using the dev build and a webpack dev server.

Error: 
```
{ TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (C:\Users\Atti.DESKTOP-8LJ8HHK\Desktop\bznz\WebDev\Projekte\strapi\node_modules\knex\lib\client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (C:\Users\Atti.DESKTOP-8LJ8HHK\Desktop\bznz\WebDev\Projekte\strapi\node_modules\knex\lib\client.js:335:13)
    at Runner.ensureConnection (C:\Users\Atti.DESKTOP-8LJ8HHK\Desktop\bznz\WebDev\Projekte\strapi\node_modules\knex\lib\runner.js:224:24)
    at Runner.run (C:\Users\Atti.DESKTOP-8LJ8HHK\Desktop\bznz\WebDev\Projekte\strapi\node_modules\knex\lib\runner.js:33:41)
    at Builder.Target.then (C:\Users\Atti.DESKTOP-8LJ8HHK\Desktop\bznz\WebDev\Projekte\strapi\node_modules\knex\lib\interface.js:23:43)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5) sql: undefined, bindings: undefined }
```

**Comment by fakedob** [comment]
Hello :)

It is my very first time giving some feedback and writing here. With just a couple of words, I am using Strapi (and its great ♥) in Azure, and most (if not all) integrations with az have been done by myself. Its a hard project, but anyway... I am happy I could get my hands on this tool.

Last night I was migrating my project from 14.5 to the latest 24.1 and while doing so, I came across something that may solve your problem, but I havent (and I will not unfortunately) had time to dig in. 

I also use mysql 5.7 so that is not the case, as I also saw the same error when I created a new project.  While I was doing the migrations on by one, I saw something interesting. If I did not specify the db at first, then the app started normally. Then, because I have implemented a lot of custom logic into the User and Auth plugin controllers, I saw that there is this new line, which was not there when I first started rewriting the plugins.

[https://github.com/strapi/strapi/blob/b0c12a3912c84fbd9d42e4f8ba55bac465a936ba/packages/strapi-plugin-users-permissions/controllers/Auth.js#L331](url)

That triggers the first time, of course when you have no admin. I really have not spent much time, I just can confirm that commenting this line and when I finally completed the migrations, the error is now gone. Funny thing is that I just tested without DB after seeing the error and I saw your post, so the 403 mentioned in the other issue is what I also saw.

Hopefully I will have more time in the future and will be able to support more on this project. Good luck

**Comment by rojer95** [comment]
I am having a similar issue
mysql version is 5.7.22

It display the following information  when I first time run strapi start
```
error (hook:bookshelf) takes too long to load
```

It display the following information when I second time run strapi start
```
UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection.
 .....
```

and then i try run strapi start again
it run success

but the api of response  status code is 403 in the admin panel

by the way, mysql is run in local (docker)

**Comment by hzburki** [comment]
I have a similar issue. My database is MySQL, hosted on AWS RDS. The project is successfully connected to the database when the new project is created.  

### Error Log
`strapi start
(node:761) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:33:41)
    at SchemaBuilder.Target.then (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/interface.js:23:43)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 6)
(node:761) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:761) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:33:41)
    at SchemaBuilder.Target.then (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/interface.js:23:43)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 7)
(node:761) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:33:41)
    at SchemaBuilder.Target.then (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/interface.js:23:43)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 18)
(node:761) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:33:41)
    at SchemaBuilder.Target.then (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/interface.js:23:43)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 19)
(node:761) UnhandledPromiseRejectionWarning: TimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at _bluebird.default.try.tap.catch (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:336:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Client_MySQL.acquireConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/client.js:335:13)
    at Runner.ensureConnection (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:224:24)
    at Runner.run (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/runner.js:33:41)
    at SchemaBuilder.Target.then (/Applications/MAMP/htdocs/numu/numu-cms/node_modules/knex/lib/interface.js:23:43)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 20)
[2019-03-20T11:12:54.642Z] debug ⛔️ Server wasn't able to start properly.
[2019-03-20T11:12:54.644Z] error (hook:bookshelf) takes too long to load
(hook:bookshelf) takes too long to load`

**Comment by hzburki** [comment]
I added some "options" to the database config file in strapi. MySQL version is **8.0.11**.

Options
```
"options": {
    "pool":{
        "min": 0,
        "max": 10,
        "idleTimeoutMillis": 3000,
        "createTimeoutMillis": 3000,
        "acquireTimeoutMillis": 3000
    }
}
```

It changed the ^ above error to
```
debug ⛔️ Server wasn't able to start properly.
error (hook:bookshelf) takes too long to load
(hook:bookshelf) takes too long to load
```

**Comment by rojer95** [comment]
> I added some "options" to the database config file in strapi. MySQL version is **8.0.11**.
> 
> Options
> 
> ```
> "options": {
>     "pool":{
>         "min": 0,
>         "max": 10,
>         "idleTimeoutMillis": 3000,
>         "createTimeoutMillis": 3000,
>         "acquireTimeoutMillis": 3000
>     }
> }
> ```
> It changed the ^ above error to
> 
> ```
> debug ⛔️ Server wasn't able to start properly.
> error (hook:bookshelf) takes too long to load
> (hook:bookshelf) takes too long to load
> ```

i try set `max` 50, then it work.

**Comment by hzburki** [comment]
I have a development database, which never exceeds more than 10 connections. But I got it to work. Refer to my last comment.

**Comment by mjurincic** [comment]
Is this a knex issue why acquireTimeoutMillis needs to be 30 seconds (30000 miliceconds) isn't that to long for databse to wait to get a connection from pool? Any resources what should be common value for that.

**Comment by jlnmay** [comment]
According to the documentation 
"acquireConnectionTimeout defaults to 60000ms and is used to determine how long knex should wait before throwing a timeout error when acquiring a connection is not possible." 

I had the same issue and set it to another value 
 acquireConnectionTimeout: process.env.KNEX_ACQUIRE_CONNECTION_TIME_OUT

**Comment by MatheusRV** [comment]
Edit database.json with this options fix for me. 

```
"options": {
        "debug": true,
        "pool": {
          "min": 0,
          "max": 10,
          "idleTimeoutMillis": 30000,
          "createTimeoutMillis": 30000,
          "acquireTimeoutMillis": 30000
        }
      }
```

**Comment by toastyghost** [comment]
Bumping timeouts to 30s and connection pool to 50 worked for me. Thanks!

**Comment by shimondoodkin** [comment]
I had a similar message with SQLite

in an event of on-save of a model I tried to execute another query

I guess the reason for my error was that it is starting a transaction when doing a save until it is saved. and strapi-connector-bookshelf is not taking pool size configuration for sqlite.

in node_modules/strapi-connector-bookshelf/lib/knex.js
about line 120 there is:
there is no way to set pool size for sqlite
 
if (connection.settings.client !== 'sqlite3') ",
     
i could replace it with if true and set {pool:{min, max...}}

or patch the default pool size in sqlite driver that is {min:1,max:1}

increasing the pool size worked well for me 

but I was afraid that it will make concurrency errors. like data saved but not reflected in results. i don't know sqlite much maybe there is no reason to worry

another option was to do set timeout of 50 ms to continue the operation after transaction completition.

**Comment by iaremarkus** [comment]
I'm getting this error on a fresh install with `npx create-strapi-app` on EC2 instance, trying to connect to postgres on RDS :(

**Comment by raudinm** [comment]
This error still persists in version 4

**Comment by nitintejuja** [comment]
im facing issue in my local pg database having components 250+ . can anybody please help me

**Comment by derrickmehaffy** [comment]
Something we noticed with another customer yesterday was an undocumented timeout setting from Knex.js. I've submitted the PR to the documentation that might help some of you.

PR: https://github.com/strapi/documentation/pull/929
Temp Doc link: https://documentation-git-fix-missingtimeoutdb-strapijs.vercel.app/developer-docs/latest/setup-deployment-guides/configurations/required/databases.html#connection-configuration-object

This doesn't "fix" the problem but acts more like a band-aid as the default will drop any connection that takes longer than 60 seconds meaning there is still an issue with some of your environments on why something would take longer than 60 seconds but increasing this timeout can help give it some more time to finish whatever it was doing.

Basically you apply it like this:

```js
module.exports = ({ env }) => ({
  connection: {
    client: 'whatever you are using',
    connection: {
      // ...
    },
    debug: false,
    acquireConnectionTimeout: 300000, // 5 minute timeout
  },
});
```
